Address logging test failures

@glatterf42 noticed test failures such as [these](https://github.com/iiasa/message-ix-models/actions/runs/8716589903/job/23914560827#step:9:10078) occurring on scheduled runs, starting today.

Symptoms:

- In general, the Pytest `caplog` fixture fails to capture certain messages, so the test assertions that check for those messages end up failing.
- Failures include:
  - Sometimes only util/test_cache.py::test_cached
  - Sometimes a longer list of 13 tests. When this occurs, they seem to all occur on the same pytest-xdist worker such as "gw1".
- All *-latest-py3.12-upstream-* jobs, both main and 3.8.0.

Actions to take:
- @khaeru to open a debugging PR and investigate if recent changes (esp #156) interfere destructively with pytest and/or recent pyam/ixmp4 releases.
- If necessary, we open a mitigation PR to skip or xfail some or all of these tests so that scheduled and PR CI can pass.

I've tried the following:

(1) Run `watchexec -e py -- "pytest -k 'test_report or test_cache' -xv -rA"`. This runs 4 tests:

- PASSED test_report.py::test_register
- PASSED test_report.py::test_report_bare_res
- PASSED test_report.py::test_report_deprecated
- FAILED test_report.py::test_report_legacy

(2) Add a debugging function in test_report.py, roughly like this:
```py
def show_log_handlers(id: int):
    import logging

    from icecream import ic

    root_logger = logging.getLogger("root")

    ic(id, root_logger.level, root_logger.handlers)
```

(3) Added calls to this function in each of the 4 tests: at least at the top and bottom, or maybe in more places.

This gives the following results (abbreviated):
```
_____________________________________________ test_register ______________________________________________
------------------------------------------ Captured stderr call ------------------------------------------
ic| id: 'test_register 0'                                                                                 
    root_logger.level: 10                                                                                 
    root_logger.handlers: [<QueueHandler (NOTSET)>,                                                       
                           <_LiveLoggingNullHandler (NOTSET)>,                                            
                           <_FileHandler /dev/null (NOTSET)>,                                             
                           <LogCaptureHandler (NOTSET)>,     
                           <LogCaptureHandler (NOTSET)>]                                                  
ic| id: 'test_register 1'                                                                                 
    root_logger.level: 10                                                                                 
    root_logger.handlers: [<QueueHandler (NOTSET)>,                                                                                  <_LiveLoggingNullHandler (NOTSET)>,                                            
                           <_FileHandler /dev/null (NOTSET)>,                                             
                           <LogCaptureHandler (NOTSET)>,                                                  
                           <LogCaptureHandler (NOTSET)>]                                                  
__________________________________________ test_report_bare_res __________________________________________
------------------------------------------ Captured stderr call ------------------------------------------
ic| id: 'test_report_bare_res 0'
    root_logger.level: 10
    root_logger.handlers: [<QueueHandler (NOTSET)>,
                           <_LiveLoggingNullHandler (NOTSET)>,
                           <_FileHandler /dev/null (NOTSET)>,
                           <LogCaptureHandler (NOTSET)>,
                           <LogCaptureHandler (NOTSET)>]
ic| id: 'test_report_bare_res 1'
    root_logger.level: 0
    root_logger.handlers: []
_________________________________________ test_report_deprecated _________________________________________
------------------------------------------ Captured stderr call ------------------------------------------
ic| id: 'test_report_deprecated 0'
    root_logger.level: 0
    root_logger.handlers: [<LogCaptureHandler (NOTSET)>, <LogCaptureHandler (NOTSET)>]
ic| id: 'test_report_deprecated 1'
    root_logger.level: 0
    root_logger.handlers: [<LogCaptureHandler (NOTSET)>, <LogCaptureHandler (NOTSET)>]
___________________________________________ test_report_legacy ___________________________________________
------------------------------------------ Captured stderr call --------------------------------
ic| id: 0
    root_logger.level: 0
    root_logger.handlers: [<LogCaptureHandler (NOTSET)>, <LogCaptureHandler (NOTSET)>]
ic| id: 1
    root_logger.level: 0
    root_logger.handlers: [<LogCaptureHandler (NOTSET)>, <LogCaptureHandler (NOTSET)>]
ic| id: 2
    root_logger.level: 0
    root_logger.handlers: [<LogCaptureHandler (NOTSET)>, <LogCaptureHandler (NOTSET)>]
```

So this tells me it is somewhere between the points "test_report_bare_res 0" and "test_report_bare_res 1" where the message-ix-models handlers get wiped out.

I'll continue to debug along these lines.

--- a/.github/workflows/pytest.yaml
+++ b/.github/workflows/pytest.yaml
@@ -103,10 +103,6 @@ jobs:
       # for open PRs), temporarily uncomment, add, or edit lines below
       # as needed. DO NOT merge such changes to `main`.
       run: |
-        # Temporary, to work around iiasa/message-ix-models#173
-        # TODO Remove after a pyam release that fixes logging
-        pip install --upgrade "pyam-iamc @ git+https://github.com/IAMconsortium/pyam.git@main"
-
         # pip install --upgrade "genno @ git+https://github.com/khaeru/genno.git@main"
         pip install --upgrade "ixmp @ git+https://github.com/iiasa/ixmp.git@${{ matrix.upstream.version }}"
         pip install --upgrade "message-ix @ git+https://github.com/iiasa/message_ix.git@${{ matrix.upstream.version }}"

